### PR TITLE
 Add cache-aware streaming and RNN-T head for Parakeet

### PR DESCRIPTION
**Pull request by jmayank1511**
# What does this PR do?

  Extends Parakeet TDT (#44171) with cache-aware streaming + an RNN-T
   head. The cache-aware variant is shipped as **Nemotron Speech
  Streaming** (`nvidia/nemotron-speech-streaming-en-0.6b`).

  ### Adds
  - `ParakeetForRNNT` — RNN-T head (LSTM prediction net + joint
  network) on the shared encoder. Supports regular and cache-aware
  checkpoints. Offline via `generate(...)`.
  - `ParakeetCacheAwareStreamingBuffer` + `streaming_step()` +
  `get_initial_streaming_state()` — chunk-by-chunk greedy RNN-T
  decoding with encoder KV cache, decoder LSTM state, and last-token
  threaded across chunks. Mirrors NeMo's
  `_greedy_decode_blank_as_pad_loop_frames`.
  - Prompt-conditioned multilingual — one-hot language MLP after
  encoder, resolved via `target_lang`.
  - NeMo `CausalConv2D` padding alignment in subsampling.
  - Unigram SentencePiece support in `ParakeetConverter`.
  - Doc + unit tests for `ParakeetForRNNT`.

## Code Agent Policy
- [x]  I confirm that this is not a pure code agent PR.

 ## Before submitting
  - [ ] This PR fixes a typo or improves the docs.
  - [x] Did you read the [contributor
  guideline](https://github.com/huggingface/transformers/blob/main/CON
  TRIBUTING.md#create-a-pull-request)?
  - [ ] Was this discussed/approved via a Github issue or the forum?
  - [x] Did you make sure to update the documentation with your
  changes?
  - [x] Did you write any new necessary tests?

## Who can review?

 @eustlb @ebezzam @vasqu

### CODE REVIEW

File: tests/models/parakeet/test_modeling_parakeet.py [hunk 1]
```diff
@@ -628,6 +630,226 @@ def test_sdpa_can_dispatch_composite_models(self):
                         raise ValueError("The eager model should not have SDPA attention layers")
 
 
+class ParakeetForRNNTModelTester:
+    def __init__(
+        self,
+        parent,
```
Comment (eustlb) [inline tests/models/parakeet/test_modeling_parakeet.py:636]: Let's add integration tests

File: src/transformers/models/parakeet/modular_parakeet.py [hunk 2]
```diff
@@ -164,6 +294,19 @@ def forward(
...
+            cache_len = cache_last_channel.shape[1]
+            cache_shape = (batch_size, cache_len, -1, self.head_dim)
+            k_cache = self.k_proj(cache_last_channel).view(cache_shape).transpose(1, 2)
+            v_cache = self.v_proj(cache_last_channel).view(cache_shape).transpose(1, 2)
+            key_states = torch.cat([k_cache, key_states], dim=2)
+            value_states = torch.cat([v_cache, value_states], dim=2)
+            new_cache = torch.cat([cache_last_channel, hidden_states], dim=1)[:, -cache_len:]
+
+        total_key_length = key_states.shape[2]
+
```
Comment (eustlb) [inline src/transformers/models/parakeet/modular_parakeet.py]: We have `SlidingWindowCache`. It should be initialised like [here](https://github.com/huggingface/transformers/blob/8d974fb67429da250a63fe6ce3cab3b4efc61656/src/transformers/models/voxtral_realtime/modular_voxtral_realtime.py#L803-L849)

then all this should be:
```suggestion
        if past_key_values is not None:
            key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx)
```

File: src/transformers/models/parakeet/modular_parakeet.py [hunk 3]
```diff
@@ -154,8 +283,9 @@ def forward(
         hidden_states: torch.Tensor,
         position_embeddings: torch.Tensor | None,
         attention_mask: torch.Tensor | None = None,
+        cache_last_channel: torch.Tensor | None = None,
         **kwargs: Unpack[TransformersKwargs],
-    ) -> tuple[torch.Tensor, torch.Tensor]:
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor | None]:
```
Comment (eustlb) [inline src/transformers/models/parakeet/modular_parakeet.py]: `past_key_values` is used throughout the lib with a strong API. Use 
```suggestion
def forward(
self,
hidden_states: torch.Tensor,
position_embeddings: torch.Tensor | None,
attention_mask: torch.Tensor | None = None,
       past_key_values: Cache | None = None,
**kwargs: Unpack[TransformersKwargs],
    ) -> tuple[torch.Tensor, torch.Tensor | None]::
```

File: src/transformers/models/parakeet/modular_parakeet.py [hunk 4]
```diff
@@ -61,9 +62,44 @@ class ParakeetEncoderModelOutput(BaseModelOutputWithPooling):
 
         - 1 for tokens that are **not masked**,
         - 0 for tokens that are **masked**.
+    cache_last_channel (`torch.Tensor` of shape `(num_layers, batch, left_ctx, hidden_size)`, *optional*):
+        Updated attention cache from the encoder (sliding KV window). Pass to the next chunk call.
+    cache_last_time (`torch.Tensor` of shape `(num_layers, batch, hidden_size, conv_left_ctx)`, *optional*):
+        Updated convolution cache from the encoder. Pass to the next chunk call.
+    cache_last_channel_len (`torch.Tensor` of shape `(batch,)`, *optional*):
+        Number of valid frames currently stored in `cache_last_channel`.
```
Comment (eustlb) [inline src/transformers/models/parakeet/modular_parakeet.py]: cf other comments but:
- kv cache should be past_key_values, as throughout the lib
- our convention for the conv cache is padding_cache

File: src/transformers/models/parakeet/modeling_parakeet.py [hunk 5]
```diff
@@ -1041,4 +1403,568 @@ def forward(
...
+        self.activation = ACT2FN[config.hidden_act]
+        self.head = nn.Linear(config.joint_hidden_size, config.vocab_size)
+
+    def forward(
+        self,
+        decoder_hidden_states: torch.Tensor,
+        encoder_hidden_states: torch.Tensor,
+    ) -> torch.Tensor:
+        joint_output = self.activation(encoder_hidden_states + decoder_hidden_states)
+        return self.head(joint_output)
```
Comment (eustlb) [inline src/transformers/models/parakeet/modeling_parakeet.py]: no reason to be defined distincly from TDTJointNetwork no?

File: src/transformers/models/parakeet/modeling_parakeet.py [hunk 6]
```diff
@@ -1041,4 +1403,568 @@ def forward(
...
+        # back into encoder_hidden. Mirrors NeMo's hybrid_rnnt_ctc_bpe_models_prompt structure.
+        self.num_prompts = config.num_prompts
+        self.prompt_dictionary = dict(config.prompt_dictionary) if config.prompt_dictionary else None
+        if self.num_prompts > 0:
+            enc_hidden = config.encoder_config.hidden_size
+            self.prompt_kernel = nn.Sequential(
+                nn.Linear(self.num_prompts + enc_hidden, 2 * enc_hidden),
+                nn.ReLU(),
+                nn.Linear(2 * enc_hidden, enc_hidden),
+            )
```
Comment (eustlb) [inline src/transformers/models/parakeet/modeling_parakeet.py:1488]: what released models use this feature? Looks like it can be removed no?

File: src/transformers/models/parakeet/modeling_parakeet.py [hunk 7]
```diff
@@ -1041,4 +1403,568 @@ def forward(
...
+        attention_mask: torch.Tensor | None = None,
+        decoder_input_ids: torch.LongTensor | None = None,
+        decoder_cache: ParakeetTDTDecoderCache | None = None,
+        use_decoder_cache: bool | None = None,
+        encoder_outputs: ParakeetEncoderModelOutput | tuple[torch.FloatTensor] | None = None,
+        labels: torch.Tensor | None = None,
+        target_lang: str | None = None,
+        prompt_id: int | None = None,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> ParakeetRNNTOutput:
```
Comment (eustlb) [inline src/transformers/models/parakeet/modeling_parakeet.py:1553]: here the signature should be (almost) the same as for voxtral realtime, specifically with `padding_cache` and `past_key_values`

**Review by eustlb — COMMENTED** [review]
Thanks a lot for raising this PR! 🤗

Looks to me that a lot of the duplicated logic from RNNT and TDT does not have to be duplicated as TDT should be an extension of RNNT

File: tests/models/parakeet/test_modeling_parakeet.py (hunk 1 above)
Comment (jmayank1511) [reply tests/models/parakeet/test_modeling_parakeet.py:636]: The model `nvidia/nemotron-3.5-asr-streaming-0.6b` will be available soon, will add the integration tests soon

File: src/transformers/models/parakeet/modular_parakeet.py (hunk 2 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modular_parakeet.py]: Done

File: src/transformers/models/parakeet/modular_parakeet.py (hunk 3 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modular_parakeet.py]: Done..

File: src/transformers/models/parakeet/modular_parakeet.py (hunk 4 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modular_parakeet.py]: Done

File: src/transformers/models/parakeet/modeling_parakeet.py (hunk 5 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modeling_parakeet.py]: Done

File: src/transformers/models/parakeet/modeling_parakeet.py (hunk 6 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modeling_parakeet.py:1488]: `nvidia/nemotron-3.5-asr-streaming-0.6b` will use this feature. This will be available in hugginface in next few days.

File: src/transformers/models/parakeet/modeling_parakeet.py (hunk 7 above)
Comment (jmayank1511) [reply src/transformers/models/parakeet/modeling_parakeet.py:1553]: Done.

**Comment by eustlb** [comment]
This was added in #46331, closing here
